apps/src/bin/quiche-server.rs

```diff
@@ -65,37 +68,61 @@ fn main() {
     let conn_args = CommonArgs::with_docopt(&docopt);
     let args = ServerArgs::with_docopt(&docopt);
 
+    let preferred_address: SocketAddr = args
+        .preferred_address
+        .parse()
+        .expect("Unable to parse preferred address.");
+
     // Setup the event loop.
     let mut poll = mio::Poll::new().unwrap();
     let mut events = mio::Events::with_capacity(1024);
 
     // Create the UDP listening socket, and register it with the event loop.
-    let mut socket =
+    let mut starting_socket =
         mio::net::UdpSocket::bind(args.listen.parse().unwrap()).unwrap();
 
-    // Set SO_TXTIME socket option on the listening UDP socket for pacing
-    // outgoing packets.
-    if !args.disable_pacing {
-        match set_txtime_sockopt(&socket) {
-            Ok(_) => {
-                pacing = true;
-                debug!("successfully set SO_TXTIME socket option");
-            },
-            Err(e) => debug!("setsockopt failed {:?}", e),
-        };
+    // Create another preferred UDP listening socket that can be used for server
+    // migration.
+    let mut preferred_socket =
+        mio::net::UdpSocket::bind(preferred_address).unwrap();
+
+    let sockets = vec![&starting_socket, &preferred_socket];
+
+    for socket in sockets {
+        // Set SO_TXTIME socket option on the listening UDP socket for pacing
+        // outgoing packets.
+        if !args.disable_pacing {
+            match set_txtime_sockopt(socket) {
+                Ok(_) => {
+                    pacing = true;
+                    debug!("successfully set SO_TXTIME socket option");
+                },
+                Err(e) => debug!("setsockopt failed {:?}", e),
+            };
+        }
     }
 
-    info!("listening on {:}", socket.local_addr().unwrap());
+    info!("listening on {:}", starting_socket.local_addr().unwrap());
+    info!("listening on {:}", preferred_socket.local_addr().unwrap());
```

Comment:
may as well update the log to state the kind of each socket

Reply:
Done 👍